### PR TITLE
improve glsl pre-roll triangles

### PR DESCRIPTION
**Pull request by m0dB**
draw the glsl preroll and postroll markers with a patterned texture and some minor additional cleanups

fixes https://github.com/mixxxdj/mixxx/issues/12015 , initially bringing GLSL waveforms in line with the legacy waveforms, but after discussion going for a slightly different look (closer to the legacy waveform look but with a semi-transparent filling)

before:
Screenshot 2023-10-13 at 19 49 49

initial after
Screenshot 2023-10-13 at 19 48 36

final after (with this PR)
Screenshot 2023-10-15 at 17 27 56

### CODE REVIEW

**Review by Swiftb0y — COMMENTED** [review]
looks good so far, no complaints from me, except that this is just the shader code atm with the code for the pre-roll still missing.

**Comment by m0dB** [comment]
oops! pushed!

**Comment by ronso0** [comment]
could your please add before / after images to the description?

**Comment by m0dB** [comment]
> could your please add before / after images to the description?

done!

(i personally like the filled rectangles better btw)

**Comment by ronso0** [comment]
> (i personally like the filled rectangles better btw)

Me too. The translucent fill is a nice compromise :+1:

File: src/waveform/renderers/allshader/waveformrendererpreroll.cpp [hunk 1]
```diff
@@ -101,45 +127,98 @@ void WaveformRendererPreroll::paintGL() {
...
+    const float imageW = static_cast<float>(imagePixelW) / devicePixelRatio;
+    const float imageH = static_cast<float>(imagePixelH) / devicePixelRatio;
+
+    QImage image(imagePixelW, imagePixelH, QImage::Format_ARGB32_Premultiplied);
+    image.setDevicePixelRatio(devicePixelRatio);
+
+    const float penWidth = 1.5f;
+    const float offset = penWidth / 2.f;
+
+    image.fill(QColor(0, 0, 0, 0).rgba());
```
Comment (Swiftb0y) [inline src/waveform/renderers/allshader/waveformrendererpreroll.cpp:194]: Does this work? Should avoid a temporary QColor construction.
```suggestion
    image.fill(Qt::transparent);
```

**Review by Swiftb0y — CHANGES_REQUESTED** [review]
Thank you. code looks in good shape, mostly minor suggestions. I really like the look of the translucent anti-aliased triangles. Reminds me of the Vital synth.

File: src/shaders/patternshader.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+varying highp vec2 vTexcoor;
+void main()
+{
+    vTexcoor = texcoor;
+    gl_Position = matrix * position;
+}
+)--");
+
+    QString fragmentShaderCode = QStringLiteral(R"--(
+uniform sampler2D sampler;
```
[the remark this answers is not in the document]
Comment (m0dB) [reply src/shaders/patternshader.cpp:19]: actually, I think I prefer ```sampler2D texture```, here and elsewhere.

[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/shaders/patternshader.cpp:19]: also works. though in this case this is the pattern shader, so I'd argue it makes more sense. But anything is better than `sampler` imo.

File: src/shaders/patternshader.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#include "shaders/patternshader.h"
+
+using namespace mixxx;
+
+void PatternShader::init() {
+    QString vertexShaderCode = QStringLiteral(R"--(
+uniform highp mat4 matrix;
+attribute highp vec4 position;
```
[the remark this answers is not in the document]
Comment (m0dB) [reply src/shaders/patternshader.cpp:8]: yes, because it is being multiplied with a mat4... which indeed could also be a mat2. i will do some additional cleanup of the shaders, while at it.

File: src/waveform/renderers/allshader/waveformrendererpreroll.cpp [hunk 4]
```diff
@@ -101,45 +127,98 @@ void WaveformRendererPreroll::paintGL() {
...
 
     glDrawArrays(GL_TRIANGLES, 0, m_vertices.size());
+}
 
-    m_shader.disableAttributeArray(vertexLocation);
-    m_shader.release();
+void WaveformRendererPreroll::generateTexture(float markerLength, float markerBreadth) {
+    const float devicePixelRatio = m_waveformRenderer->getDevicePixelRatio();
+    m_markerLength = markerLength;
+    m_markerBreadth = markerBreadth;
```
[the remark this answers is not in the document]
Comment (m0dB) [reply src/waveform/renderers/allshader/waveformrendererpreroll.cpp:181]: agreed, will change

File: src/shaders/patternshader.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (m0dB) [reply src/shaders/patternshader.cpp:8]: of course the day we want waveforms projected on rotating 3d cubes will will have to revert this ;-)

File: src/waveform/renderers/allshader/waveformrendererpreroll.h [hunk 5]
```diff
@@ -17,15 +19,23 @@ class WaveformRendererPreroll;
...
+    void drawPattern(float x1, float y1, float x2, float y2, float repetitions, bool flip);
+    void generateTexture(float markerLength, float markerBreadth);
+
+    mixxx::PatternShader m_shader;
     QColor m_color;
+    float m_markerBreadth{};
+    float m_markerLength{};
+    std::unique_ptr<QOpenGLTexture> m_pTexture;
     VertexData m_vertices;
+    VertexData m_texcoords;
```
[the remark this answers is not in the document]
Comment (m0dB) [reply src/waveform/renderers/allshader/waveformrendererpreroll.h:38]: the only reason was to use the VertexData::addRectangle to populate the vertex values, and making them members it to reuse the memory (which is not particularly relevant here, but in line with the other renderers).

not very strong reasons, so I will change this.

File: src/shaders/patternshader.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (m0dB) [reply src/shaders/patternshader.cpp:8]: i will do a bit of type cleanup though

File: src/waveform/renderers/allshader/waveformrendererpreroll.cpp (hunk 1 above)
Comment (m0dB) [reply src/waveform/renderers/allshader/waveformrendererpreroll.cpp]: that works

**Comment by m0dB** [comment]
@Swiftb0y I have addressed all comments, please mark the conversations as resolved when re-reviewed.

File: src/shaders/patternshader.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/shaders/patternshader.cpp:8]: > there is no advantage in doing the conversion from vec2 to vec4 manually, and doing it like this keeps the shader more flexible (usuable for 3d, which is of course not our usecase, but since we get it for free, why cripple the shader?)

Sure, makes sense. the vec3 is probably padded to 4 anyways. This just needs a comment then.

File: src/waveform/renderers/allshader/waveformrendererpreroll.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/waveform/renderers/allshader/waveformrendererpreroll.h]: Being able to use `VertexData::addRectangle` is a valid reason. Reusing the memory is also valid but unnecessary if the data could just be stack-local.

File: src/widget/wspinnyglsl.cpp [hunk 6]
```diff
@@ -97,17 +97,17 @@ void WSpinnyGLSL::paintGL() {
     m_textureShader.bind();
 
     int matrixLocation = m_textureShader.matrixLocation();
-    int samplerLocation = m_textureShader.samplerLocation();
+    int textureLocation = m_textureShader.textureLocation();
     int positionLocation = m_textureShader.positionLocation();
     int texcoordLocation = m_textureShader.texcoordLocation();
 
-    QMatrix4x4 matrix;
+    QMatrix matrix;
```
Comment (Swiftb0y) [inline src/widget/wspinnyglsl.cpp:104]: `QMatrix` is deprecated in Qt5 and removed in Qt6, why did you change to it?

File: src/waveform/renderers/allshader/waveformrendererpreroll.cpp [hunk 7]
```diff
@@ -12,6 +12,53 @@
 #include "widget/wskincolor.h"
 #include "widget/wwidget.h"
 
+namespace {
+QOpenGLTexture* generateTexture(float markerLength,
```
Comment (Swiftb0y) [inline src/waveform/renderers/allshader/waveformrendererpreroll.cpp:16]: I'm sorry for expanding the scope yet again, but this needs a little more supporting refactoring. The pointer returned here (and thus by createTexture) is owning. We should really avoid those in new code. Can you refactor this and `createTexture` to return by value or `unique_ptr` (preferably the former).

**Comment by Swiftb0y** [comment]
> please mark the conversations as resolved when re-reviewed.

Did so and also reviewed the new code. Thank you.

File: src/widget/wspinnyglsl.cpp (hunk 6 above)
Comment (m0dB) [reply src/widget/wspinnyglsl.cpp:104]: good question. i'll change it back.

File: src/waveform/renderers/allshader/waveformrendererpreroll.cpp (hunk 7 above)
Comment (m0dB) [reply src/waveform/renderers/allshader/waveformrendererpreroll.cpp:16]: sure. i will use a unique_ptr though

Comment (Swiftb0y) [reply src/waveform/renderers/allshader/waveformrendererpreroll.cpp]: Sure... seems a little unnecessary, but I won't insist.

File: src/shaders/vinylqualityshader.cpp [hunk 8]
```diff
@@ -5,7 +5,7 @@ using namespace mixxx;
 void VinylQualityShader::init() {
     QString vertexShaderCode = QStringLiteral(R"--(
 uniform mat4 matrix;
-attribute highp vec4 position;
+attribute highp vec4 position; // use vec4 here (will be padded) for matrix multiplication
 attribute highp vec2 texcoord;
 varying highp vec2 vTexcoord;
 void main()
```
Comment (Swiftb0y) [inline src/shaders/vinylqualityshader.cpp:11]: we should really deduplicate some of these shaders, but thats out-of-scope.

File: src/util/texture.cpp [hunk 9]
```diff
@@ -1,31 +1,31 @@
 #include "util/texture.h"
 
-QOpenGLTexture* createTexture(const QImage& image) {
+std::unique_ptr<QOpenGLTexture> createTexture(const QImage& image) {
     if (image.isNull()) {
-        return nullptr;
+        return std::unique_ptr<QOpenGLTexture>{};
```
Comment (Swiftb0y) [inline src/util/texture.cpp:5]: why do you prefer this over just returning `nullptr`?

File: src/util/texture.cpp [hunk 10]
```diff
@@ -1,31 +1,31 @@
...
     }
-    QOpenGLTexture* pTexture = new QOpenGLTexture(image);
+    std::unique_ptr<QOpenGLTexture> pTexture{new QOpenGLTexture(image)};
```
Comment (Swiftb0y) [inline src/util/texture.cpp:7]: We use `make_unique` everywhere. https://stackoverflow.com/a/22571331
```suggestion
    auto pTexture = std::make_unique<QOpenGLTexture>(image);
```

**Review by Swiftb0y — COMMENTED** [review]
Thank you, small nitpicks.

File: src/util/texture.cpp (hunk 9 above)
Comment (m0dB) [reply src/util/texture.cpp:5]: I like to be explicit, but I'm okay with changing that if you commonly return nullptr elsewhere.

File: src/util/texture.cpp (hunk 10 above)
Comment (m0dB) [reply src/util/texture.cpp]: Thought we were restricted to c++11

Comment (Swiftb0y) [reply src/util/texture.cpp]: Fortunately not. 2.4 is restricted to gcc10, so C++17 is fully okay. On main, you can even use C++20.

File: src/util/texture.cpp (hunk 9 above)
Comment (Swiftb0y) [reply src/util/texture.cpp:5]: I'd prefer that.

Comment (m0dB) [reply src/util/texture.cpp]: ok, changed back. i think with that I addressed all your comments.

**Review by Swiftb0y — APPROVED** [review]
Yes, LGTM. Thank you very much. Highly appreciated. The anti-aliasing is a huuge improvement. Any objections @Holzhaus?
CI failures is unrelated, IIUC @daschuer is looking into it.

File: src/widget/wspinnyglsl.cpp [hunk 11]
```diff
@@ -138,15 +138,15 @@ void WSpinnyGLSL::paintGL() {
...
+        rotate.rotate(-m_fGhostAngle, 0, 0);
         m_textureShader.setUniformValue(matrixLocation, rotate);
 
         drawTexture(m_pGhostTextureScaled.get());
     }
 
     if (m_pFgTextureScaled) {
         QMatrix4x4 rotate;
-        rotate.rotate(m_fAngle, 0, 0, -1);
+        rotate.rotate(-m_fAngle, 0, 0);
```
Comment (JoergAtGithub) [inline src/widget/wspinnyglsl.cpp:149]:
```suggestion
        rotate.rotate(m_fAngle, 0, 0, -1);
```
Same here

**Review by JoergAtGithub — CHANGES_REQUESTED** [review]

**Comment by JoergAtGithub** [comment]
I just tested this PR with vinyl control signal and the spinnies show it without color (just black):
grafik

**Comment by m0dB** [comment]
@JoergAtGithub Aargh, the only think I forgot to test myself! Thanks, I will fix this now.

**Comment by m0dB** [comment]
@JoergAtGithub I have fixed the vinyl quality shader. If you confirm it works, this is ready for merge.

**Comment by Holzhaus** [comment]
> Yes, LGTM. Thank you very much. Highly appreciated. The anti-aliasing is a huuge improvement. Any objections @Holzhaus?

Looks nice. My only objection regarding the previous look was that Software/GL/GLSL/Shader waveform rendering should be consistent. Is this the case now.

**Comment by m0dB** [comment]
With 50% transparency, it is 50% more consistent ;-)

At least it is now clearly deliberate and can't be mistaken for a bug. So IMO it is "consistent" enough and sacrificing a bit of consistency for a nicer look makes sense. 

We could try to change the legacy waveforms to match this look, but I don't think that's worth the effort, particularly taking into account that there is already a lot of variation between the different waveform types.

**Comment by JoergAtGithub** [comment]
The color of the Vinylcontrol signal is now back. 

I'm not sure if the following is a regression, or was always the case, but when I drop the needle, I see the signal outside of the round area:

https://github.com/mixxxdj/mixxx/assets/64457745/f7a308f9-c952-41e5-a869-a45cd5894365

Maybe it's just me, but I expected this to be masked as the Cover-Art.

**Comment by m0dB** [comment]
> Maybe it's just me, but I expected this to be masked as the Cover-Art.

That would be as simple as swapping the draw order of the mask and the vinyl quality. But I took this order from the original spinny code, and changing this should not be part of this PR.

Merge?

**Comment by JoergAtGithub** [comment]
Since it's not a regression, let's merge this as it is!

**Comment by Swiftb0y** [comment]
Thank for this nice visual improvement @m0dB. Thank you for the thorough testing @JoergAtGithub.
